rir/src/compiler/translations/pir_2_rir/allocators.h

```diff
@@ -451,26 +451,88 @@ class CachePosition {
   public:
     typedef size_t SlotNumber;
     typedef std::pair<SEXP, Value*> NameAndEnv;
+    typedef std::pair<size_t, size_t> StartSize;
 
-    size_t size() { return uniqueNumbers.size(); }
+    explicit CachePosition(Code* code) {
+        std::unordered_map<Value*,
+                           std::unordered_map<NameAndEnv, size_t, pairhash>>
+            found;
 
-    SlotNumber isCached(const NameAndEnv& key) {
-        // Those are stored directly in symbols
-        if (Env::Cast(key.second) &&
-            (Env::Cast(key.second)->rho == R_BaseEnv ||
-             Env::Cast(key.second)->rho == R_BaseNamespace))
-            return false;
-        if (uniqueNumbers.size() == MAX_CACHE_SIZE)
-            return uniqueNumbers.count(key);
-        return true;
+        // As a simple heuristic, we cache all variable which are used more
+        // than once.
+        Visitor::run(code->entry, [&](Instruction* i) {
+            if (auto ld = LdVar::Cast(i)) {
+                found[ld->env()][NameAndEnv(ld->varName, ld->env())]++;
+            } else if (auto st = StVar::Cast(i)) {
+                found[st->env()][NameAndEnv(st->varName, st->env())]++;
+            }
+        });
+
+        // First all global envs
+        for (const auto& env : found) {
+            auto e = Env::Cast(env.first);
+            if (e && e->rho != R_BaseEnv && e->rho != R_BaseNamespace) {
+                for (const auto& key : env.second) {
+                    if (key.second <= 2)
+                        continue;
+                    if (uniqueNumbers.size() == MAX_CACHE_SIZE)
+                        break;
+                    uniqueNumbers.emplace(key.first, uniqueNumbers.size());
+                }
+            }
+        }
+        globalEnvsCacheSize_ = uniqueNumbers.size();
+
+        // Heuristic: after how many loads does it pay off to cache a variable?
+        // It depends on the size of the environment, since the smaller the
+        // environment, the faster the lookup.
```

Comment:
Why the smaller the faster? It is a hash table...
Oth, the code defines bigger limits for smaller environment
And then, why for non-global environments we cache even if there is only one load while for global we do not?
Also how the heuristic treat loop lds is unclear...

Reply:
> Why the smaller the faster? It is a hash table...

no, its a list. lookup is a linear search

> Oth, the code defines bigger limits for smaller environment

yes, the smaller the env, the more lookups are needed to win anything by caching

> And then, why for non-global environments we cache even if there is only one load while for 
> global we do not?

well if you think about it, for local envs the number of bindings we statically see is roughly equal to the size of the environment. but for a global env, that is certainly not true, as the global env will have many more unrelated bindings. so there this heuristic does not work.

> Also how the heuristic treat loop lds is unclear...

well it doesn't, but in practice it still seems to work quite nicely

Reply:
I didn't get it. You were talking about the standard lookup. I understood it as the binding cache lookup.
I would add to the comment ...the smaller the env, the faster the (standard|noncached) lookup. 
Oth, this is not taking into account that the lookup could need to go through a hierarchy of envs.